### PR TITLE
[Hub Generated] Review request for Microsoft.Devices to add version stable/2023-06-30

### PR DESCRIPTION
**Pull request by MikalaiSavenka**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/MikalaiSavenka-iothub-Microsoft.Devices-2023-06-30...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [ ] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by TimLovellSmith** [comment]
>         },

are these create/update only?

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3255 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         },

Having name, subscriptionId, and resourceGroup, and id seems redundant - are they all identifying the same resource? Or are there diffeent resources here?

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3225 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "Id of the cosmos DB sql container endpoint",

Want to include an example?

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3215 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         },

format: uri?

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3229 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "The template for generating a synthetic partition key value for use with this cosmos DB sql container. The template must include at least one of the following placeholders: {iothub}, {deviceid}, {DD}, {MM}, and {YYYY}. Any one placeholder may be specified at most once, but order and non-placeholder components are arbitrary. This parameter is only required if PartitionKeyName is specified.",

Why is it required if PartitionKeyName is specified?

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3269 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "The user assigned identity."

objectid of?

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:4485 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         "cosmosDBSqlContainers": {

Are you going to add an example for this? It'd be nice if your examples cover all the different kind of routing endpoints with a common set of parameters.

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:2952 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         },

not sure I understand question

---
In reply to: [1542561780](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542561780) [](http://example.com/codeflow?ancestors=1542561780,1542561780)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3255 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         },

what is annotating as a resource id? do you have any documentation link?

---
In reply to: [1542565092](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542565092) [](http://example.com/codeflow?ancestors=1542565092,1542565092)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3217 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         },

name - name of endpoint set by customer
id  - internal id
resourceGroup - it is azure resource group where cosmosdb account reside

---
In reply to: [1542562663](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542562663) [](http://example.com/codeflow?ancestors=1542562663,1542562663)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3225 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         },

the example is: [https://ailn-systemstore-test2.documents.azure.com:443/](https://ailn-systemstore-test2.documents.azure.com:443/)

I dont see we use format in this doc at all

---
In reply to: [1542565328](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542565328) [](http://example.com/codeflow?ancestors=1542565328,1542565328)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3229 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>           "description": "The template for generating a synthetic partition key value for use with this cosmos DB sql container. The template must include at least one of the following placeholders: {iothub}, {deviceid}, {DD}, {MM}, and {YYYY}. Any one placeholder may be specified at most once, but order and non-placeholder components are arbitrary. This parameter is only required if PartitionKeyName is specified.",

partitionkeyname is name stamped to document
partitinkeytemplate is value of stamped document

---
In reply to: [1542566830](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542566830) [](http://example.com/codeflow?ancestors=1542566830,1542566830)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3269 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         "cosmosDBSqlContainers": {

I will add example for cosmosdb endpoint, as far as this PR is for cosmosdb endpoint

---
In reply to: [1542569674](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542569674) [](http://example.com/codeflow?ancestors=1542569674,1542569674)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:2952 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         "cosmosDBSqlContainers": {

added

---
In reply to: [1542569674](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1542569674) [](http://example.com/codeflow?ancestors=1542569674,1542569674)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:2952 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>         },

So now, looks like you're using x-ms-secret, which implies x-ms-mutability is create/update only. I think its probably good to make that explicit though.

Do I understand right that you allow the customer to pass the keys they like, to enable iothub to connect to their cosmosDB instance?

This sounds like a pretty painful way of doing it, for anyone who wants to or has to (compliance) implementi key autorotations. Firstly, they can't even tell/audit what key values are in use.

---
In reply to: [1593816631](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1593816631) [](http://example.com/codeflow?ancestors=1593816631,1542561780)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3255 in 09336e3. [](commit_id = 09336e3a5ddf329ae15f5c593921f5cc06086eb8, deletion_comment = False)

File: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/examples/iothub_addRoutingCosmosDBEndpoint.json [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+          },
+          "routing": {
+            "endpoints": {
+              "serviceBusQueues": [],
+              "serviceBusTopics": [],
+              "eventHubs": [],
+              "storageContainers": [],
+              "cosmosDBSqlContainers": [
+                {
+                  "id": "009208fa-9cb7-4b90-bfbd-f40086e766c9",
```
Comment (TimLovellSmith) [inline specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/examples/iothub_addRoutingCosmosDBEndpoint.json:163]: >  "id": "009208fa-9cb7-4b90-bfbd-f40086e766c9", [](http://example.com/codeflow?start=16&length=47)

In ARM APIs 'id' often means the resource id. I suppose you have to do it this way for consistency, but if not, I would have wondered if we could find a better name, maybe uuid or uniqueId, or if it is even really needed. Is there any way the customer will consume this by passing it to other parts of the API?

**Comment by TimLovellSmith** [comment]
>         },

Is either databaseName or name, specifically the name of the ARM cosmosDB resource? Trying to understand intended usage.

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3261 in 13514e2. [](commit_id = 13514e28bd3f4c8c3236a093234a7af885448028, deletion_comment = False)

**Comment by MikalaiSavenka** [comment]
>         },

name of database created within cosmosdb resource

---
In reply to: [1612439265](https://github.com/Azure/azure-rest-api-specs/pull/23791#issuecomment-1612439265) [](http://example.com/codeflow?ancestors=1612439265,1612439265)

---
Refers to: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/iothub.json:3261 in 13514e2. [](commit_id = 13514e28bd3f4c8c3236a093234a7af885448028, deletion_comment = False)

File: specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/examples/iothub_addRoutingCosmosDBEndpoint.json (hunk 1 above)
Comment (MikalaiSavenka) [reply specification/iothub/resource-manager/Microsoft.Devices/stable/2023-06-30/examples/iothub_addRoutingCosmosDBEndpoint.json:163]: id is inherited name as in all other endpoints

**Comment by TimLovellSmith** [comment]
Added DoNotMerge at requet of PR author

**Comment by TimLovellSmith** [comment]
Merge conflicts need resolving now..

**Comment by TimLovellSmith** [comment]
@rkmanda I wonder why powershell fails. Seems like its something to do wtih the tags in the readme maybe? Its a bit beyond me.

>WARNING: /mnt/vss/_work/1/s/azure-rest-api-specs/specification
04:26:26.784 cmdout 	[psci.ps1] WARNING: There should be only one tag  in /mnt/vss/_work/1/s/azure-rest-api-specs/specification/marketplacecatalog/resource-manager/readme.md
04:26:26.792 cmdout 	[psci.ps1] WARNING: There should be only one tag  in /mnt/vss/_work/1/s/azure-rest-api-specs/specification/marketplacecatalog/resource-manager/Solutions/readme.md
04:26:28.003 cmdout 	[psci.ps1] WARNING: There should be only one tag  in /mnt/vss/_work/1/s/azure-rest-api-specs/specification/machinelearning/resource-manager/readme.md

**Review by TimLovellSmith — APPROVED** [review]

**Review by rkmanda — APPROVED** [review]

**Pull request merged**